Add kubestash docs (#1)
kubestash/docs: docs/concepts/crds/snapshot/index.md

    identifier: snapshot-overview
    name: Snapshot
    parent: crds
    weight: 55
product_name: kubestash
menu_name: docs_{{ .version }}
section_menu_id: concepts
---
> New to KubeStash? Please start [here](/docs/concepts/README.md).

# Snapshot

## What is Snapshot

> A `Snapshot` is a Kubernetes `CustomResourceDefinition`(CRD) which represents the state of a backup run to a particular `Repository`. Multiple components of the same target may be backed up in the same `Snapshot`. This is a namespaced CRD. It should be in the same namespace as the respective `Repository`. 

Reviewer: ```suggestion
A `Snapshot` is a Kubernetes `CustomResourceDefinition`(CRD) which represents the state of a backup run for one or more components of an application.
```

> KubeStash operator is responsible for creating `Snapshot` custom resource. `Snapshot` is not supposed to be created/edited by the end user.

Reviewer: A `Snapshot` is a Kubernetes `CustomResourceDefinition`(CRD) which represents the state of a backup run for one or more components of an application. For every `BackupSession`, it creates `Snapshot` CRs. If a `BackupSession` involves multiple repositories, a `Snapshot` is created  for each repository.

**spec.backupSession** 

`spec.backupSession` represents the name of the respective `BackupSession` which is responsible for this `Snapshot`.

**spec.version**

`spec.version` denotes the respective data organization structure inside the `Repository`.

**spec.appRef**

`spec.appRef` specifies the reference of the application that has been backed up in this `Snapshot`.

**spec.deletionPolicy**

`spec.deletionPolicy` specifies what to do when you delete a `Snapshot` CR. The valid values are:
- **Delete** This will delete just the Snapshot CR from the cluster but keep the backed up data in the remote backend. This is the default behavior.
- **WipeOut** This will delete the Snapshot CR as well as the backed up data from the backend.

**spec.paused**

`spec.paused` specifies whether the `Snapshot` is paused or not. If the `Snapshot` is paused, KubeStash will not process any further event for the `Snapshot`.

### Snapshot `Status`

`Snapshot` object has the following fields in `.status` section:

**status.phase**

`status.phase` represents the backup state of this `Snapshot`.

**status.snapshotTime**

> `status.snapshotTime` represents the timestamp when this `Snapshot` was taken.

Reviewer: ```suggestion
`status.snapshotTime` represents the original creation timestamp for this `Snapshot`.
```

**status.lastUpdateTime**

`status.lastUpdateTime` specifies the timestamp when this Snapshot was last updated.

**status.size**

`status.size` represents the size of the `Snapshot`.

**status.integrity**

`status.integrity` represents whether the `Snapshot` data has been corrupted or not.

**status.conditions**

`status.conditions` represents list of conditions regarding this `Snapshot`. KubeStash sets the following conditions for a `Snapshot`:

| Field                       | Usage                                                          |

> |-----------------------------|----------------------------------------------------------------|
> | `SnapshotMetadataUploaded`  | Indicates whether the metadata of Snapshot is uploaded or not. |
> | `RecentSnapshotListUpdated` | Indicates whether the recent Snapshot list is updated or not.  |
> 
> **status.totalComponents**

Reviewer: ```suggestion
|-----------------------------|----------------------------------------------------------------|
| `SnapshotMetadataUploaded`  | Indicates whether the metadata of Snapshot was uploaded or not. |
| `RecentSnapshotListUpdated` | Indicates whether the recent Snapshot list was updated or not.  |

**status.totalComponents**
```

- **phase :** represents the backup phase of the component.
- **size :** represents the size of the restic repository for this component.
- **duration :** specifies the total time taken to complete the backup process for this component.
- **integrity :** represents the result of the restic repository integrity check for this component.
- **error :** specifies the reason in case of backup failure for the component.
- **driver :** specifies the name of the tool that has been used to upload the underlying backed up data.
- **resticStats :** specifies the **Restic** driver specific information. Each resticStat consists of the following fields:
  - **id :** represents the restic snapshot id.
  - **uploaded :** specifies the amount of data that has been uploaded in the restic snapshot.
  - **hostPath :** represents the backup path for which restic snapshot is taken.
  - **size :** represents the restic snapshot size.
- **volumeSnapshotterStats :** specifies the **VolumeSnapshotter** driver specific information. Each volumeSnapshotterStat consists of the following fields:

>   - **pvcName :** represents the backup PVC name for which volumeSnapshot is created.
>   - **hostPath :** represents the corresponding path of PVC for which volumeSnapshot is created.

Reviewer: ```suggestion
  - **pvcName :** represents the PVC name for which volumeSnapshot was created.
  - **hostPath :** represents the mount path of PVC for which volumeSnapshot was created.
```
